projects/rero/ng-core/src/lib/record/editor/utils.ts

```diff
@@ -14,7 +14,102 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
+import { FormlyFieldConfig } from '@ngx-formly/core';
 import { extractIdOnRef } from '../../utils/utils';
+import { ILogger } from '../../ILogger';
+
+/**
+ * Convert form to widget definition in JSONSchema
+ * @param schema - object, the JSONSchema
+ * @param logger - Logger
+ * @returns object, a converted JSONSchema
+ */
+export function formToWidget(schema: any, logger: ILogger): any {
+  if (schema.properties) {
+    for (const property of Object.keys(schema.properties)) {
+      formToWidget(schema.properties[property], logger);
+    }
+  }
+
+  /* items */
+  if (schema.items) {
+    formToWidget(schema.items, logger);
+  }
+
+  /* oneOf, anyOf, allOf */
+  ['oneOf', 'anyOf', 'allOf'].map((type: string) => {
+    if (schema[type]) {
+      schema[type].map((element: any) => formToWidget(element, logger));
+    }
+  });
+
+  /* form (custom) */
+  if (schema.form) {
+    /* Initialize the widget key on JSONSchema */
+    if (!schema.widget) {
+      schema.widget = {};
+    }
+    if (!schema.widget.formlyConfig) {
+      schema.widget.formlyConfig = {};
+    }
+
+    /* Template options */
+    if (schema.form.templateOptions) {
+      templateOptionsCreateIfNotExist(schema.widget.formlyConfig);
+      Object.keys(schema.form.templateOptions).map((key: string) => {
+        switch (key) {
+          case 'wrappers':
+            schema.widget.formlyConfig.wrappers = schema.form.templateOptions[key];
+            break;
+          default:
+            schema.widget.formlyConfig.templateOptions[key] = schema.form.templateOptions[key];
+        }
+      });
+      if (Object.keys(schema.widget.formlyConfig.templateOptions).length === 0) {
+        delete schema.widget.formlyConfig.templateOptions;
+      }
+      delete schema.form.templateOptions;
+    }
+
+    [
+      'hide', 'focus', 'expressionProperties', 'hideExpression', 'type'
+    ].map((option: string) => {
+      if (option in schema.form) {
+        schema.widget.formlyConfig[option] = schema.form[option];
+        delete schema.form[option];
+      }
+    });
+
+    [
+      'helpURL', 'fieldMap', 'remoteOptions', 'selectWithSortOptions',
+      'remoteTypeahead', 'validation', 'navigation', 'placeholder', 'options'
+    ].map((option: string) => {
+      if (option in schema.form) {
+        templateOptionsCreateIfNotExist(schema.widget.formlyConfig);
+        schema.widget.formlyConfig.templateOptions[option] = schema.form[option];
+        delete schema.form[option];
+      }
+    });
+
+    if (Object.keys(schema.form).length !== 0) {
+      logger.error(schema.form, 'The form key definition in JSONSCHEMA is not empty.');
```

Comment:
Isn't it necessary to raise an error at this point ? Why continue the processing if an error occurred ?